backend/controller/cronjobs/cronjobs.go

```diff
@@ -153,344 +71,140 @@ func (s *Service) NewCronJobsForModule(ctx context.Context, module *schemapb.Mod
-func (s *Service) UpdatedControllerList(ctx context.Context, controllers []parentdal.Controller) {
+// scheduleCronJob schedules the next execution of a single cron job.
+func (s *Service) scheduleCronJob(ctx context.Context, tx *Tx, job model.CronJob) error {
 	logger := log.FromContext(ctx).Scope("cron")
-	controllerIdx := -1
-	for idx, controller := range controllers {
-		if controller.Key.String() == s.key.String() {
-			controllerIdx = idx
-			break
-		}
+	now := s.clock.Now().UTC()
+	pending, err := tx.db.IsCronJobPending(ctx, job.Key, now)
+	if err != nil {
+		return fmt.Errorf("failed to check if cron job %q is pending: %w", job.Key, err)
 	}
-	if controllerIdx == -1 {
-		logger.Tracef("controller %q not found in list of controllers", s.key)
+	if pending {
+		logger.Tracef("Attempt to schedule cron job %q which is already pending", job.Key)
+		return nil
 	}
 
-	oldState := s.hashRingState.Load()
-	if oldState != nil && len(oldState.controllers) == len(controllers) {
-		hasChanged := false
-		for idx, new := range controllers {
-			old := oldState.controllers[idx]
-			if new.Key.String() != old.Key.String() {
-				hasChanged = true
-				break
-			}
-		}
-		if !hasChanged {
-			return
-		}
-	}
-
-	hashRing := hashring.New(slices.Map(controllers, func(c parentdal.Controller) string { return c.Key.String() }))
-	s.hashRingState.Store(&hashRingState{
-		hashRing:    hashRing,
-		controllers: controllers,
-		idx:         controllerIdx,
-	})
-
-	s.events.Publish(updatedHashRingEvent{})
-}
-
-// isResponsibleForJob indicates whether a this service should be responsible for attempting jobs,
-// or if enough other controllers will handle it. This allows us to spread the job load across controllers.
-func (s *Service) isResponsibleForJob(job model.CronJob, state *state) bool {
-	if state.isJobTooNewForHashRing(job) {
-		return true
+	pattern, err := cron.Parse(job.Schedule)
+	if err != nil {
+		return fmt.Errorf("failed to parse cron schedule %q: %w", job.Schedule, err)
 	}
-	hashringState := s.hashRingState.Load()
-	if hashringState == nil {
-		return true
+	originTime := job.StartTime
+	if t, ok := job.LastExecution.Get(); ok {
+		originTime = t
 	}
 
-	initialKey, ok := hashringState.hashRing.GetNode(job.Key.String())
-	if !ok {
-		return true
+	nextAttemptForJob, err := cron.NextAfter(pattern, originTime, false)
+	if err != nil {
+		return fmt.Errorf("failed to calculate next execution for cron job %q with schedule %q: %w", job.Key, job.Schedule, err)
+	}
+	if nextAttemptForJob.Before(now) {
+		nextAttemptForJob = now
 	}
 
-	initialIdx := -1
-	for idx, controller := range hashringState.controllers {
-		if controller.Key.String() == initialKey {
-			initialIdx = idx
-			break
-		}
+	logger.Tracef("Scheduling cron job %q async_call execution at %s", job.Key, nextAttemptForJob)
+	refKey := schema.RefKey{Module: job.Verb.Module, Name: job.Verb.Name}
+	origin := fmt.Sprintf("cron:%s", job.Key)
```

Comment:
I would strongly type this with the correct origin type, and use its `String()` method to give you the string representation where needed. This will make it easier to catch any potential changes to the underlying type structure.